File: src/features/typing/data/go/valid_parentheses.go
package main

func isValid(s string) bool {
	stack := make([]byte, 0, len(s))
	pairs := map[byte]byte{
		')': '(',
		']': '[',
		'}': '{',
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c == '(' || c == '[' || c == '{' {
			stack = append(stack, c)
			continue
		}
		n := len(stack)
		if n == 0 || stack[n-1] != pairs[c] {
			return false
		}
		stack = stack[:n-1]
	}
	return len(stack) == 0
}

File: src/features/typing/data/ts/valid_parentheses.ts
function isValid(s: string): boolean {
  const stack: string[] = [];
  const pairs: Record<string, string> = {
    ')': '(',
    ']': '[',
    '}': '{'
  };
  for (const c of s) {
    if (c === '(' || c === '[' || c === '{') {
      stack.push(c);
      continue;
    }
    const top = stack.pop();
    if (top !== pairs[c]) {
      return false;
    }
  }
  return stack.length === 0;
}
